File: 11495 Bubbles and Buckets/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void MergeSort(int *array, int *reg, int left, int right, long *swap) {
    if(left < right) {
        int mid = left + (right - left) / 2;
        MergeSort(array, reg, left, mid, swap);
        MergeSort(array, reg, mid + 1, right, swap);
        int leftPointer = left, rightPointer = mid + 1;
        for(int i = left; i <= right; ++i) {
            if(leftPointer > mid) {
                reg[i] = array[rightPointer++];
                *swap += mid - leftPointer + 1;
            } else if(rightPointer > right)
                reg[i] = array[leftPointer++];
            else if(array[leftPointer] <= array[rightPointer])
                reg[i] = array[leftPointer++];
            else {
                reg[i] = array[rightPointer++];
                *swap += mid - leftPointer + 1;
            }
        }
        for(int i = left; i <= right; ++i)
            array[i] = reg[i];
    }
}

int main() {
    int N;
    char *buffer = NULL;
    size_t bufsize = 0;
    while(scanf("%d", &N) && N != 0) {
        getline(&buffer, &bufsize, stdin);
        char *token = strtok(buffer, " ");
        int P[N], reg[N];
        for(int i = 0; i < N; ++i) {
            P[i] = atoi(token);
            token = strtok(NULL, " ");
        }
        long swap = 0;
        MergeSort(P, reg, 0, N - 1, &swap);
        puts(swap % 2 == 1 ? "Marcelo" : "Carlos");
    }
    free(buffer);
	return 0;
}

File: 11495 Bubbles and Buckets/solution.java
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		int N;
		StringBuilder output = new StringBuilder();
		while(st.nextToken() == StreamTokenizer.TT_NUMBER && (N = (int)st.nval) != 0) {
		    int[] P = new int[N], reg = new int[N];
		    for(int i = 0; i < N; ++i) {
		        st.nextToken();
		        P[i] = (int)st.nval;
		    }
		    long[] swap = new long[1];
		    MergeSort(P, reg, 0, N - 1, swap);
		    output.append((swap[0] & 1) == 1 ? "Marcelo\n" : "Carlos\n");
		}
		System.out.print(output);
	}
	
	static void MergeSort(int[] P, int[] reg, int left, int right, long[] swap) {
        if(left < right) {
            int mid = left + ((right - left) >> 1);
            MergeSort(P, reg, left, mid, swap);
            MergeSort(P, reg, mid + 1, right, swap);
            int leftPointer = left, rightPointer = mid + 1;
            for(int i = left; i <= right; ++i) {
                if(leftPointer > mid) {
                    reg[i] = P[rightPointer++];
                    swap[0] += mid - leftPointer + 1;
                } else if(rightPointer > right)
                    reg[i] = P[leftPointer++];
                else if(P[leftPointer] <= P[rightPointer])
                    reg[i] = P[leftPointer++];
                else {
                    reg[i] = P[rightPointer++];
                    swap[0] += mid - leftPointer + 1;
                }
            }
            System.arraycopy(reg, left, P, left, right - left + 1);
        }
    }
}
